### PR TITLE
Fix `cwd` issues when using Neomake

### LINKED ISSUE #260: Automatic rustfmt erases my files sometimes.
* rust.vim version: Current one, I use vim-plug and do `:PlugUpdate` often.

Steps to reproduce:

I think it started happening when I converted my project to use a workspace. The current state is: https://github.com/dpc/crev/commit/d9fbf06bb1385181b3788fc0cffd544ec66ff565

Expected vs. actual behavior:

Typically my vim consistently reformats a file on every save.

Now, for whatever reason, it stops working in the middle of the session for some files (for which it worked fine just a second ago) and from that time it just keeps erasing them every time I write the buffer. :/

Paste debugging info from the Rust Vim plugin via _one_ of the following
commands: `:RustInfo`,  `:RustInfoToClipboard`, or `:RustInfoToFile <filename>`.

```
rust.vim Global Variables:

let g:ftplugin_rust_source_path = v:null
let g:loaded_syntastic_rust_cargo_checker = v:null
let g:loaded_syntastic_rust_filetype = v:null
let g:loaded_syntastic_rust_rustc_checker = v:null
let g:rust_bang_comment_leader = v:null
let g:rust_cargo_avoid_whole_workspace = v:null
let g:rust_clip_command = v:null
let g:rust_conceal = v:null
let g:rust_conceal_mod_path = v:null
let g:rust_conceal_pub = v:null
let g:rust_fold = v:null
let g:rust_last_args = v:null
let b:rust_last_args = []
let g:rust_last_rustc_args = v:null
let b:rust_last_rustc_args = []
let g:rust_original_delimitMate_excluded_regions = v:null
let g:rust_playpen_url = v:null
let g:rust_prev_delimitMate_quotes = v:null
let g:rust_recent_nearest_cargo_tol = v:null
let g:rust_recent_root_cargo_toml = v:null
let g:rust_recommended_style = v:null
let g:rust_set_conceallevel = v:null
let g:rust_set_conceallevel=1 = v:null
let g:rust_set_foldmethod = v:null
let g:rust_set_foldmethod=1 = v:null
let g:rust_shortener_url = v:null
let g:rustc_makeprg_no_percent = v:null
let g:rustc_path = v:null
let g:rustfmt_autosave = 0
let g:rustfmt_autosave_because_of_config = v:null
let g:rustfmt_autosave_if_config_present = 1
let g:rustfmt_command = 'rustfmt'
let g:rustfmt_emit_files = 1
let g:rustfmt_fail_silently = 0
let g:rustfmt_options = ''
let g:syntastic_extra_filetypes = ['rust']
let g:syntastic_rust_cargo_fname = v:null
rustfmt 0.99.4-nightly (1c40881 2018-08-27)

rustc 1.30.0-nightly (fc81e3624 2018-09-07)

cargo 1.29.0-nightly (0ec7281b9 2018-08-20)


NVIM v0.3.1
Build type: Release
LuaJIT 2.1.0-beta3
Compilation: /nix/store/wkfafgwgs1na5f808wybs8l6v8yjrvqz-gcc-wrapper-7.3.0/bin/gcc -Wconversion -O2 -DNDEBUG -DMIN_LOG_LEVEL=3 -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wimplicit-fallthrough -Wvla -fstack-protector-strong -fdiagnostics-color=auto -Wno-array-bounds -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -I/build/source/build/config -I/build/source/src -I/nix/store/3jzza70h4xlcl65py98y0ddq3sr3nv6z-libuv-1.21.0/include -I/nix/store/wqmabb91mn7c8l0cbagbi7n6b78r2k6g-libmsgpack-2.1.5/include -I/nix/store/qwlgay0pdgllmdwif144771z8y0anx3j-unibilium-2.0.0/include -I/nix/store/caygb8bjk106ja8fddlx2zw5fbzzz8g2-libtermkey-0.20/include -I/nix/store/r54qm8js4ndph7mqwm61fgpd54niwbxi-neovim-libvterm-2017-11-05/include -I/nix/store/k8fcymm135rp923s2w8w5z7wzr2kdcvi-jemalloc-5.0.1/include -I/nix/store/z10kkr081zb02kyjw8ybf2s9zll49cvx-glibc-2.27-dev/include -I/build/source/build/src/nvim/auto -I/build/source/build/include
Compiled by nixbld

Features: +acl +iconv +jemalloc +tui 
See ":help feature-compile"

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/nix/store/19sgqgn1vpc8c7828hhf0lmpaqx9yb8m-neovim-unwrapped-0.3.1/share/nvim"

Run :checkhealth for more info
```

### PR DESCRIPTION
**Pull request by dpc**
All credit to @blueyed .

This fixes #260

The investigation is in the corresponding Neomake issue:
https://github.com/neomake/neomake/issues/2080

This is also related to: #259

### CODE REVIEW

File: autoload/rustfmt.vim [hunk 1]
```diff
@@ -146,8 +146,8 @@ function! s:RunRustfmt(command, tmpname, fail_silently)
             let l:content = readfile(a:tmpname)
         endif
 
-        call writefile(l:content, expand('%'))
-        silent edit!
+        normal! ggdG
+        call setline(1, l:content)
         let &syntax = &syntax
```
Comment (blueyed) [inline autoload/rustfmt.vim]: The `let &syntax = &syntax` is also not necessary anymore likely (if it ever was, but maybe due to the `edit` that was used previously).

TLDR of the investigation: the problem is that `expand('%')` here might be relative, but it was `:lcd`ed before.
An alternative might be to use the full buffer name (save it before `cd`ing), but I think just setting the lines is better than writing from outside.

**Comment by blueyed** [comment]
btw: as can be seen in https://github.com/neomake/neomake/issues/2080#issuecomment-419692234 the error/issue shows up immediately when not using `:silent!` - therefore this should really be avoided, since it hides all kind of errors.

File: autoload/rustfmt.vim [hunk 2]
```diff
@@ -146,8 +146,8 @@ function! s:RunRustfmt(command, tmpname, fail_silently)
             let l:content = readfile(a:tmpname)
         endif
 
-        call writefile(l:content, expand('%'))
-        silent edit!
+        normal! ggdG
```
Comment (blueyed) [inline autoload/rustfmt.vim]: Just noticed that this should become `normal! "_ggdG` to not mess with registers.

File: autoload/rustfmt.vim (hunk 1 above)
Comment (blueyed) [reply autoload/rustfmt.vim]: Actually `expand('%')` should be the relative path after `:lcd` - so there is some other reason for this (maybe a bug in Vim/Neovim even (NVIM v0.3.1 used by the reporter)).

Anyway, the `cd`ing back should be done in a `try … finally` block, because otherwise it might not change it back.

**Comment by blueyed** [comment]
Please remove the mentioning of me before merging it, otherwise it will show up in notifications all over the place.  Use `@blueyed` maybe in backticks or so.

**Comment by blueyed** [comment]
Thanks for trying - I was still notified of the new commits though.. :/

**Comment by da-x** [comment]
Hi @blueyed and @dpc ,

Thanks for debugging the issue.

However the problem with the `normal! "_ggdG` and `call setline(1, l:content)` sequence is that it's a regression over issue #236. We should also prevent undo actions from being affected by this fix.

**Comment by blueyed** [comment]
@da-x 
I see (https://github.com/rust-lang/rust.vim/commit/2fb3df5c53770d7710d0171a6d5ea7767f86bce9).
I've looked at neoformat, and they use a similar approach: https://github.com/sbdchd/neoformat/blob/master/autoload/neoformat.vim#L170
Maybe it is more related to the (missing/wrong) undojoin, and or `mkview!` etc?

**Comment by blueyed** [comment]
@dpc 
Not sure.  I think it still is a bit fragile what is being done in rust.vim, but there appear to be reasons, and I am not using it myself anyway.

I suggest closing this for now - it can get re-opened later.

**Comment by da-x** [comment]
@dpc @blueyed Looking over neoformat, I pushed 4c21791115d44e to master. It should do something similar this PR, but without regressing over #236. The existing `undojoin` seems to work with this change without any cursor funkiness on `u`. I wished though that `winsaveview` and `winrestview` would work right like in neoformat but was not able to achieve that. Feel free to test.

**Comment by blueyed** [comment]
@da-x 
Looks good, thanks!

**Comment by dpc** [comment]
I've updated all vim plugins to latest and will keep working on my workspace-enabled project. I'll report any issues. Thanks everyone for looking into this!
